### ISSUE TITLE
User-defined temperatures of collision partners

### ISSUE THREAD

**Issue by migueldvb** [lime-rt/lime#173]
It is not supported in the code to have a different temperature distribution for each collisional partner. This could be useful to be able to define a temperature profile for the electrons when they are included as collisional partner and they are thermally decoupled from the neutral gas.

**Comment by tlunttil** [comment]
It shouldn't be too difficult to allow the possibility of several temperature distributions. However, before writing a PR it's maybe better to wait until we've decided whether we're going to make changes to some of the related data structures. I think @allegroLeiden  may have some plans about that.

**Comment by allegroLeiden** [comment]
Yes, I think it would be better to wait at least until after 1.7 before thinking about enhancements. The code structure ought to settle down by then.

**Comment by migueldvb** [comment]
I think what is needed for the case of collisional partners with different temperature distribution is to be able to define collision rates as a function of position in the grid instead of reading them directly from the LAMDA datafiles. It should be possible to include a function in the model file that overrides the collision rates defined in the molecular datafile and at the same time maintaining backward compatibility.

**Comment by tlunttil** [comment]
A problem with using a user-defined collision rate function is that it can cause a severe slowdown, because the function would be called many times when the stateq. system is constructed during iteration. With careless use of transcendental functions and `pow()`s that could easily become the main time sink.

One solution would be to precalculate the rate coefficients during grid construction and store them (in fact LIME used to do something like that), but that approach can easily lead to huge memory needs in case large models and molecules with many collisional transitions.

**Comment by migueldvb** [comment]
Those are very good points. For some problems it may be necessary to have collision rates that depend on the position or some parameter other than the gas temperature, so it owuld be useful to have this functionality with a warning to users about possible problems.

**Comment by tlunttil** [comment]
That should be easy to do, because the only place that needs to call the coll. coeffs function seems to be line 148 of stateq.c. If the change is done, it would probably also be a good idea to move the computation of the collision part of the stateq. matrix outside the ALI loop. That would reduce the number of calls to the coll. coeffs function by at least a factor of 5.

**Comment by allegroLeiden** [comment]
Michiel Hogerheijde: Using different temperatures per collision partner is a must. However, I don't understand the argument about collision rates not being calculated for each grid point: they already have to be, since only the downward rates are given and the upward rates are calculated (they depend on T anyway; I may have gotten up- and down- confused here).

**Comment by tlunttil** [comment]
Collisional rates have to be, and are, calculated at each grid point on the fly (precalcing during grid initialisation and storing the rates is be too memory intensive for large models and molecules with many levels). The problem is  that implementing a fully general collision rate calculation will require calling user-written functions in some of the inner loops, which has the potential for causing severe performance problems. I agree with what @migueldvb said, we could include this option and add a warning in the documentation.
